### PR TITLE
[PLAT-129]auth poc related work for validating JWT Token

### PR DESCRIPTION
**Pull request by sateeshvouch**
This is just an illustration of how we are going to validate the JWT Tokens on our API side. There are several other things on Auth0 side that are configured to get this POC working. Please refer to Readme on more details about it.

1. This repo is created from our typescript-service-template [repo](https://github.com/svinstech/typescript-service-template).
2. I tried not include the repo setup/cleanup changes in to this PR, but you will find few of them.

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -1,15 +1,19 @@
...
 
 DD_APM_ENABLED=true # Set this to false to disable datadog integration
 DD_PORT=8126
-DD_SERVICE=service_template_app
+DD_SERVICE=api_auth_poc_app
 DD_ENV=dev
 DD_LOGS_INJECTION=true
 DD_RUNTIME_METRICS_ENABLED=true
+
+# AUTH SPECIFIC CONFIG
```
Comment (sateeshvouch) [inline .env.sample:17]: Required auth specific environment configuration.

File: package.json [hunk 2]
```diff
@@ -29,12 +29,16 @@
...
+        "@nestjs/passport": "^8.0.1",
         "@nestjs/platform-express": "^8.1.2",
         "@nestjs/swagger": "^5.1.4",
         "@nestjs/typeorm": "^8.0.2",
         "class-transformer": "^0.4.0",
         "class-validator": "^0.13.1",
         "express": "^4.17.1",
+        "jwks-rsa": "^2.0.5",
+        "passport": "^0.5.0",
+        "passport-jwt": "^4.0.0",
```
Comment (sateeshvouch) [inline package.json:41]: Passport Strategy  specific to JWT workflow.

File: package.json [hunk 3]
```diff
@@ -29,12 +29,16 @@
...
         "@nestjs/core": "^8.1.2",
+        "@nestjs/passport": "^8.0.1",
         "@nestjs/platform-express": "^8.1.2",
         "@nestjs/swagger": "^5.1.4",
         "@nestjs/typeorm": "^8.0.2",
         "class-transformer": "^0.4.0",
         "class-validator": "^0.13.1",
         "express": "^4.17.1",
+        "jwks-rsa": "^2.0.5",
+        "passport": "^0.5.0",
```
Comment (sateeshvouch) [inline package.json:40]: Auth middle for NodeJS

File: package.json [hunk 4]
```diff
@@ -29,12 +29,16 @@
         "@nestjs/common": "^8.1.2",
         "@nestjs/config": "^1.1.0",
         "@nestjs/core": "^8.1.2",
+        "@nestjs/passport": "^8.0.1",
```
Comment (sateeshvouch) [inline package.json:32]: Passport utilites for Nest. [Docs](https://www.npmjs.com/package/@nestjs/passport)

File: package.json [hunk 5]
```diff
@@ -29,12 +29,16 @@
...
         "@nestjs/platform-express": "^8.1.2",
         "@nestjs/swagger": "^5.1.4",
         "@nestjs/typeorm": "^8.0.2",
         "class-transformer": "^0.4.0",
         "class-validator": "^0.13.1",
         "express": "^4.17.1",
+        "jwks-rsa": "^2.0.5",
```
Comment (sateeshvouch) [inline package.json:39]: Library to retrieve signing keys. [Docs](https://www.npmjs.com/package/jwks-rsa)

File: src/api-auth-poc/api-auth-poc.module.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+// A new Auth module.
+
+import { Module } from '@nestjs/common';
+
+import { HealthController } from './controllers/health.controller';
+import { ApplicationController } from './controllers/application.controller';
+
+@Module({
+    imports: [],
+    controllers: [HealthController, ApplicationController],
```
Comment (sateeshvouch) [inline src/api-auth-poc/api-auth-poc.module.ts:10]: A new Auth module, list of controllers that this module is applicable for.

File: src/api-auth-poc/controllers/application.controller.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
+// An application controller.
```
Comment (sateeshvouch) [inline src/api-auth-poc/controllers/application.controller.ts:1]: A new `application` controller that offers two routes.

`api/v1/application/public`
`api/v1/application/private`

File: src/api-auth-poc/controllers/application.controller.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+               } 
+            }
+        */
+        /* Include business logic to validate,
+             1. If the partner has access to this user's application backed by our DB.
+             2. If the user belongs to that company and has access to this application (Backed by our DB, not sure if this can be answered by our existing DB Model).
+        */
+        return {
+            application: 'privateApplicationId',
+            partnerKey: req.user.partnerKey,
```
Comment (sateeshvouch) [inline src/api-auth-poc/controllers/application.controller.ts:34]: This is dummy data, we can replace this response schema with actual application data.

File: src/api-auth-poc/controllers/health.controller.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import { Controller, Get, Header } from '@nestjs/common';
```
Comment (sateeshvouch) [inline src/api-auth-poc/controllers/health.controller.ts:1]: unrelated, you can ignore

File: src/api-auth-poc/entities/base_entity.entity.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import {
```
Comment (sateeshvouch) [inline src/api-auth-poc/entities/base_entity.entity.ts:1]: unrelated, you can ignore

File: src/api-auth-poc/services/type-orm-config.service.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import { Inject, Injectable } from '@nestjs/common';
```
Comment (sateeshvouch) [inline src/api-auth-poc/services/type-orm-config.service.ts:1]: unrelated, you can ignore.

File: src/app.module.ts [hunk 12]
```diff
@@ -12,10 +11,8 @@ const ENV = process.env.NODE_ENV;
             isGlobal: true,
             envFilePath: !ENV ? '.env' : `.env.${ENV}`,
         }),
-        TypeOrmModule.forRootAsync({
-            useClass: TypeOrmConfigService,
-        }),
-        ServiceTemplateModule,
+        APIAuthPOCModule,
+        AuthModule,
```
Comment (sateeshvouch) [inline src/app.module.ts:15]: Registering our Auth Module here.

File: src/auth/auth.module.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import { Module } from '@nestjs/common';
```
Comment (sateeshvouch) [inline src/auth/auth.module.ts:1]: Auth module, with auth strategies defined.

File: src/auth/jwt.strategy.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import { Injectable } from '@nestjs/common';
+import { PassportStrategy } from '@nestjs/passport';
+import { ExtractJwt, Strategy } from 'passport-jwt';
+import { passportJwtSecret } from 'jwks-rsa';
+import * as dotenv from 'dotenv';
+
+dotenv.config();
+
+const USER_CONTEXT: string = 'https://api.vouch.us/context/user';
```
Comment (sateeshvouch) [inline src/auth/jwt.strategy.ts:9]: Namespace to look for the `user` information on the access_token

File: src/auth/jwt.strategy.ts [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+const PARTNER_CONTEXT: string = 'azp';
```
Comment (sateeshvouch) [inline src/auth/jwt.strategy.ts:10]: Namespace of claim to look for `partner` information on the access_token.

File: src/auth/jwt.strategy.ts [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+            jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+            audience: process.env.AUTH0_AUDIENCE,
+            issuer: `${process.env.AUTH0_ISSUER_URL}`,
+        });
+    }
+
+    // Verifies and validates the payload.
+    validate(payload: any): any {
+        return {
```
Comment (sateeshvouch) [inline src/auth/jwt.strategy.ts:30]: Passing the user and partner context further down the request chain.

File: src/main.ts [hunk 17]
```diff
@@ -15,6 +15,10 @@ async function bootstrap() {
         .build();
     const document = SwaggerModule.createDocument(app, options);
     SwaggerModule.setup('api', app, document);
+
+    // Adding path prefix at the global level.
+    app.setGlobalPrefix('api/v1');
```
Comment (sateeshvouch) [inline src/main.ts:20]: Tweaked this config, to make the urls look nice with postman.

File: tsconfig.json [hunk 18]
```diff
@@ -22,6 +22,7 @@
         // added in Typescript 4.4 this should be true when we move up
         //"exactOptionalPropertyTypes": true,
         "noImplicitOverride": true,
-        "noImplicitReturns": true
+        "noImplicitReturns": true,
+        "strictPropertyInitialization": false
```
Comment (sateeshvouch) [inline tsconfig.json:26]: I had to turn this off, as some of the entity objects were complaining.

File: src/api-auth-poc/controllers/health.controller.spec.ts [hunk 19]
```diff
@@ -0,0 +1,16 @@
+import { getMockControllers } from '../../test/mockData';
```
Comment (sateeshvouch) [inline src/api-auth-poc/controllers/health.controller.spec.ts:1]: unrelated, you can ignore this.

File: src/api-auth-poc/controllers/application.controller.ts [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+        return { application: 'publicAplicationId', partnerKey: null };
+    }
+
+    // Private route.
+    @UseGuards(AuthGuard('jwt'))
+    @Get('/private')
+    privateResponse(@Request() req: any): any {
+        /* In the request object we now have the context of the claims like below
+            req.user = {
+                partnerKey = "kxrODb3qdrkyHMvq3fBOVxneID7OX5QW"
```
Comment (sateeshvouch) [inline src/api-auth-poc/controllers/application.controller.ts:21]: Need to tighten here on snake_case vs camelCase.

File: src/api-auth-poc/controllers/application.controller.ts [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+        /* In the request object we now have the context of partner & user like below
+            req.user = {
+                partnerKey = "kxrODb3qdrkyHMvq3fBOVxneID7OX5QW"
+                userInfo: {
+                    "user_id": "sateesh.kadiyala@vouch.us",
+                    "comapny_id": "c154bf3c-36d0-40ec-a687-abed1c2ce85a"
+               } 
+            }
+        */
+        /* Include business logic to validate,
```
Comment (samdalton) [inline src/api-auth-poc/controllers/application.controller.ts:28]: I imagine we can extract this into a shared function/library for re-use

Comment (sateeshvouch) [reply src/api-auth-poc/controllers/application.controller.ts:28]: I m thinking our `*.strategy.ts` (for example in our case `jwt.strategy.ts`) is the last stop for the authN/Z code. The strategy class would rely on an `AuthService` that talks to a corresponding DBService like Partner/User for validations.
We use those AuthService validation methods in our strategy to fwd/deny the requests here.
This pattern is described well on NestJS [docs](https://docs.nestjs.com/security/authentication#login-route).

Comment (samdalton) [reply src/api-auth-poc/controllers/application.controller.ts:28]: Makes sense!

**Review by samdalton — APPROVED** [review]

**Pull request merged**